### PR TITLE
Simulation

### PR DESCRIPTION
**Pull request by Yadunund**

### CODE REVIEW

File: src/simulation.md [hunk 1]
```diff
@@ -2,3 +2,202 @@
...
+various stages of R&D and deployment. More notably, simulations provide the
+following benefits.
+
+- **Time and resource saving:** While testing with hardware is indispensible,
+  the process can slow the pace of development with additional setup time, robot
+  downtime and and reset time periods between trials. As the number of
+  participants scale, so do costs associated with purchasing hardware and
+  consumables for testing. This is especially true with solutions such as RMF
+  which aim to integrate several mobile/stationary robots, and building systems
+  such as doors and lifts. Simulations provide a cost effective and time saving
```
Comment (mdfesto) [inline src/simulation.md]:
```suggestion
  such as doors and lifts. Simulations provide a potentially cost effective and time saving
```

File: src/simulation.md [hunk 2]
```diff
@@ -2,3 +2,202 @@
...
+  alternative for evaluating the behavior of robot systems at scale. More
+  importantly simulations can help answer questions with deployment such as how
```
Comment (mdfesto) [inline src/simulation.md]:
```suggestion
  importantly simulations can help answer questions prior to deployment such as how
```

File: src/simulation.md [hunk 3]
```diff
@@ -2,3 +2,202 @@
...
+  many participants can be supported or how the existing behavior would change
+  with the introduction of a new fleet, both of which can inform purchasing
+  decisions for facility owners. 
+
+- **Robust testing:** Robots in simulation do not run out of battery. Scenarios
```
Comment (mdfesto) [inline src/simulation.md]:
```suggestion
- **Robust testing:** Robots in simulation do not run out of battery and don't cost any money if your robots happen to unfortunately crash into something. Scenarios
```

File: src/simulation.md [hunk 4]
```diff
@@ -2,3 +2,202 @@
...
+  can be tested for hours at a stretch, at faster speeds, to fine tune
+  algorithms and verify their robustness. As scenarios in simulation are
```
Comment (mdfesto) [inline src/simulation.md]:
```suggestion
  algorithms and verify their robustness. One factor on how much scenario testing to run is a decision on how much compute power you want to avail for the simulation. With the introduction of cloud simulation, this limit is then a trade off of cost and speed as well. As scenarios in simulation are
```

File: src/simulation.md [hunk 5]
```diff
@@ -2,3 +2,202 @@
...
+  consequences can also be studied through simulation. Data logged from hardware
+  trials can be used to recreate the scenario in simulation which may be helpful
+  for debugging. Lastly, long running simulations can instill confidence in
+  facility owners.
+
+Physics-based simulators such as `Gazebo`, carry the benefit of easily
+interfacing with ROS2 nodes through wrappers provided by `gazebo_ros_pkgs`.
+Gazebo plugins can be developed that accurately emulate the behavior of robots,
+sensors and infrastructure systems which enhance the overall fidelity of
+simulations. It is worth emphasizing here that the code used to run simulations
```
Comment (mdfesto) [inline src/simulation.md]:
```suggestion
simulations. It is worth emphasizing here that the exact same code used to run the simulations
```

File: src/simulation.md [hunk 6]
```diff
@@ -2,3 +2,202 @@
...
+will be run on physical hardware as well without any changes.
```
Comment (mdfesto) [inline src/simulation.md]:
```suggestion
will also be run on the physical system as well without any changes.
```

File: src/simulation.md [hunk 7]
```diff
@@ -2,3 +2,202 @@
...
+
+However, despite these compelling benefits, simulations are sparingly employed
+by developers and system integrators citing complexity over generating
+environments and configuring them with appropriate plugins. The RMF project also
```
Comment (mdfesto) [inline src/simulation.md]:
```suggestion
environments and configuring them with appropriate plugins. In a recent publication "_A Study on the Challenges of Using Robotics Simulators for Testing_," by Afsoon Afzal, Deborah S. Katz, Claire Le Goues and Christopher S. Timperley they noted the main reasons participants gave for not using simulation for a particular project and summarized their findings as follows:

| Reason for not using simulation  | #  | %  |
|---|---|---|
| Lack of time or resources | 15 | 53.57% |
| Not realistic/accurate enough | 15 | 53.57% |
| Lack of expertise or knowledge on how to use software-based simulation | 6  | 21.43% |
| There was no simulator for the robot | 4  | 14.29% |
| Not applicable | 4  | 14.29% |
| Too much time or compute resources | 2  | 7.14%  |
| Nobody suggested it | 0  | 0.00%  |
| Other | 2  | 7.14%  |

The RMF project also
```

File: src/simulation.md [hunk 8]
```diff
@@ -2,3 +2,202 @@
...
+aims to address these hurdles by simplifying the process of setting up
+simulation environments for multi-fleet traffic control.
```
Comment (mdfesto) [inline src/simulation.md]:
```suggestion
simulation environments for multi-fleet traffic control as we will explain further throughout this section.
```

File: src/simulation.md (hunk 7 above)
Comment (mdfesto) [reply src/simulation.md]: Not sure why this table isn't formatting correctly yet...

File: src/simulation.md [hunk 9]
```diff
@@ -2,3 +2,202 @@
...
+
+## Building Map Generator
+
+
+## RMF Assets and Plugins
+
+Assets play a pivotal role in recreating environments in simulation. Over the
```
Comment (mdfesto) [inline src/simulation.md]:
```suggestion
Assets play a pivotal role in recreating environments in simulation. Fortunately there are 
```

File: src/simulation.md [hunk 10]
```diff
@@ -2,3 +2,202 @@
...
+course of the RMF project, several 3D models of robots, mechanical
```
Comment (mdfesto) [inline src/simulation.md]:
```suggestion
projects such as RMF, SubT and others which allow developers to create and open source 3D models of robots, mechanical
```

File: src/simulation.md [hunk 11]
```diff
@@ -2,3 +2,202 @@
...
+infrastructure systems and scene objects have been open sourced. They are
```
Comment (mdfesto) [inline src/simulation.md]:
```suggestion
infrastructure systems and scene objects. They are
```

File: src/simulation.md [hunk 12]
```diff
@@ -2,3 +2,202 @@
...
+core systems through the aid of plugins. 
+
+To simulate the behavior of hardware such as robot models and infrastructure
+systems(doors & lifts), several Gazebo plugins have been architected. These
+plugins are derivates of the [ModelPlugin](http://osrf-distributions.s3.amazonaws.com/gazebo/api/dev/classgazebo_1_1ModelPlugin.html)
+class and tie in standard ROS2 and rmf_core messages to provide the necessary
+functionality. The following sections briefly describe some of these plugins.
+
+#### Robots
+As highlighted earlier, several robot models (MiR100, Magni, Hospi) have been
```
Comment (mdfesto) [inline src/simulation.md]:
```suggestion
As highlighted earlier, several robot models (SESTO, MiR100, Magni, Hospi) have been
```

File: src/simulation.md [hunk 13]
```diff
@@ -2,3 +2,202 @@
...
+```
+
+The door [plugin](https://github.com/osrf/traffic_editor/blob/master/building_gazebo_plugins/src/door.cpp) responds to `DoorRequest` messages with `door_name` matching its `model name` sdf tag. These messages are published over the `/door_requests` topic. The plugin is agnostic of the type of door defined and relies on the `left_joint_name` and `right_joint_name` parameters to determine which joints to actuate during open and close motions. During these motions, the joints are commanded to their appropriate limits which are specified in the parent element. The joint motions follow similar acceleration and deceleration profiles as in the `slotcar`. 
+ 
+
+#### Lifts
+
+
+#### Dispensers
+A common scenario of interest is that of robots performing deliveries where a payload is loaded onto the robot at one location and the unloaded at another.
```
Comment (mdfesto) [inline src/simulation.md]:
```suggestion
A common scenario of interest is that of robots performing deliveries where a payload is loaded onto a robot at one location and then unloaded at another. These load and unload tasks can involve robot manipulator actions as well.
```

File: src/simulation.md [hunk 14]
```diff
@@ -2,3 +2,202 @@
...
+</model>
+```
+
+The door [plugin](https://github.com/osrf/traffic_editor/blob/master/building_gazebo_plugins/src/door.cpp) responds to `DoorRequest` messages with `door_name` matching its `model name` sdf tag. These messages are published over the `/door_requests` topic. The plugin is agnostic of the type of door defined and relies on the `left_joint_name` and `right_joint_name` parameters to determine which joints to actuate during open and close motions. During these motions, the joints are commanded to their appropriate limits which are specified in the parent element. The joint motions follow similar acceleration and deceleration profiles as in the `slotcar`. 
+ 
+
+#### Lifts
+
+
+#### Dispensers
```
Comment (mdfesto) [inline src/simulation.md:309]: Are we calling these "Dispensers" or "Workcells"?

**Review by mdfesto — COMMENTED** [review]
This is looking great! I made a few change requests for your review.

File: src/simulation.md (hunk 14 above)
Comment (Yadunund) [reply src/simulation.md:309]: I guess "Dispensers" are a class of "Workcells". Happy to modify however to be consistent with the rest of the book.

File: src/simulation.md (hunk 13 above)
Comment (Yadunund) [reply src/simulation.md]: The comment seems to be on an outdated version.

**Pull request merged**